### PR TITLE
speed up study.best_trial on RDB.

### PR DESCRIPTION
**Pull request by chris-chris**
Speed up the best trial function.
I just made PR again for some conflicts.

[Original]
https://github.com/optuna/optuna/pull/686

### CODE REVIEW

File: optuna/storages/rdb/models.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from sqlalchemy import orm
 from sqlalchemy import String
 from sqlalchemy import UniqueConstraint
+from sqlalchemy import desc, asc
```
Comment (hvy) [inline optuna/storages/rdb/models.py]: Could you import the two symbols in separate lines, for consistency with the above imports?

File: optuna/storages/rdb/models.py [hunk 2]
```diff
@@ -161,6 +162,26 @@ def find_by_id(cls, trial_id, session):
 
         return trial
 
+    @classmethod
+    def find_max_value_trial(cls, study_id, session):
+        # type: (int, orm.Session) -> TrialModel
```
Comment (hvy) [inline optuna/storages/rdb/models.py]: Could you apply code style fixes according to https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#empty-lines-after-type-hints ?

File: optuna/storages/rdb/storage.py [hunk 3]
```diff
@@ -671,6 +671,30 @@ def get_all_trials(self, study_id):
...
+
+        params = models.TrialParamModel.where_trial(trial, session)
+        values = models.TrialValueModel.where_trial(trial, session)
+        user_attributes = models.TrialUserAttributeModel.where_trial(trial, session)
+        system_attributes = models.TrialSystemAttributeModel.where_trial(trial, session)
+
+        frozen_trial = self._merge_trials_orm([trial], params, values, user_attributes,
+                                              system_attributes)[0]
+
+        self._finished_trials_cache.cache_trial_if_finished(frozen_trial)
```
Comment (hvy) [inline optuna/storages/rdb/storage.py]: Did you forget to read from the cache, in order to actually make use of it (c.f. `get_trial`)?

File: optuna/storages/rdb/models.py [hunk 4]
```diff
@@ -161,6 +162,26 @@ def find_by_id(cls, trial_id, session):
 
         return trial
 
+    @classmethod
+    def find_max_value_trial(cls, study_id, session):
```
Comment (hvy) [inline optuna/storages/rdb/models.py]: You need to also filter by state to only query completed trials I think.

**Comment by hvy** [comment]
Could you update the title of this PR to be a bit more descriptive?

**Comment by hvy** [comment]
@chris-chris Do you have any status updates of this PR? I really appreciate your contribution so far and would like to proceed with merging these changes as similar issues have been reported from other sources (#760). Let me know if you're busy and we can figure something out.

**Comment by chris-chris** [comment]
@hvy It's hard to fix the original unit test bug reports since I do not know details of your codes.
Can you help me fixing unit test reports?

**Comment by hvy** [comment]
Sorry for the delay response, I'll take a look.

**Comment by hvy** [comment]
I believe the errors will be fixed by the following change https://github.com/hvy/optuna/commit/78bff4b1d0a72ab911f88febe8783c070c9c56d5 (which by the way addresses https://github.com/optuna/optuna/pull/729#discussion_r350007199). Could you apply a similar fix?

To elaborate a bit on the failure, the failure states 
```
        ...
        best_value = self.best_trial.value
>       assert best_value is not None
E       AssertionError
```
on this line, https://github.com/optuna/optuna/blob/master/optuna/study.py#L76. This is because the returned trial could still be running (its state is `TrialState.RUNNING`) in which case its `value` will be `None`. We must filter them out in `best_trial`.

File: optuna/storages/rdb/storage.py (hunk 3 above)
Comment (chris-chris) [reply optuna/storages/rdb/storage.py]: I replaced the code to`get_trial`

File: optuna/storages/rdb/models.py (hunk 1 above)
Comment (chris-chris) [reply optuna/storages/rdb/models.py]: done

Comment (chris-chris) [reply optuna/storages/rdb/models.py]: done )

File: optuna/storages/rdb/models.py (hunk 2 above)
Comment (chris-chris) [reply optuna/storages/rdb/models.py]: done!

File: optuna/storages/rdb/storage.py (hunk 3 above)
Comment (chris-chris) [reply optuna/storages/rdb/storage.py]: done!

**Comment by chris-chris** [comment]
@hvy all tests have passed, please check )

**Comment by sile** [comment]
🎉

**Pull request merged**